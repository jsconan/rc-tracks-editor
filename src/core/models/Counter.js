/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { validateCallback } from '../helpers';
import { eventEmitterMixin } from '../mixins';

/**
 * Represents a set of counters.
 */
export class Counter extends Map {
    /**
     * Creates an observable set of counters.
     * Each change in the counters will be notified to the subscribers.
     * @param {*} source - An iterable object that can be used to initialize the counters.
     */
    constructor(source = null) {
        super();

        eventEmitterMixin(this);

        if (source) {
            this.load(source);
            source = void 0;
        }
    }

    /**
     * Applies a callback to each counter from the list.
     * @param {function} walker - A callback that will be applied to each counter of the list.
     * @returns {Counter} - Chains the instance.
     * @throws {TypeError} - If the given callback is not a function.
     */
    forEach(walker) {
        validateCallback(walker);

        super.forEach((value, key) => walker.call(this, value, key, this));

        return this;
    }

    /**
     * Maps the counters to an object.
     * @param {function} mapper - A mapper callback that will be applied to each counter.
     * @returns {object}
     * @throws {TypeError} - If the given callback is not a function.
     */
    map(mapper) {
        validateCallback(mapper);

        const counters = {};

        for (const [key, value] of this.entries()) {
            counters[key] = mapper.call(this, value, key, this);
        }

        return counters;
    }

    /**
     * Gets the value for a particular counter.
     * @param {string} key - The key of the counter to read.
     * @returns {number} - The current value of the counter. It returns 0 if the counter does not exist.
     */
    get(key) {
        return super.get(key) || 0;
    }

    /**
     * Sets the value for a particular counter.
     * @param {string} key - The key of the counter to set.
     * @param {number} value  - The new value of the counter. It must be a whole number.
     * @returns {Counter} - Chains the instance.
     * @fires set
     */
    set(key, value) {
        const previous = super.get(key) || 0;
        super.set(key, Math.floor(value) || 0);

        this.emit('set', key, value, previous);

        return this;
    }

    /**
     * Removes a counter.
     * @param {string} key - The key of the counter to delete.
     * @returns {boolean} - Returns `true` if a counter was deleted.
     * @fires delete
     */
    delete(key) {
        const value = super.get(key) || 0;
        const deleted = super.delete(key);

        if (deleted) {
            this.emit('delete', key, value);
        }

        return deleted;
    }

    /**
     * Increments a counter.
     * @param {string} key - The key of the counter to increment.
     * @param {number} amount - The amount to add to the counter.
     * @returns {Counter} - Chains the instance.
     * @fires set
     */
    increment(key, amount = 1) {
        const value = Math.floor(amount) || 1;
        return this.set(key, this.get(key) + value);
    }

    /**
     * Decrements a counter.
     * @param {string} key - The key of the counter to decrement.
     * @param {number} amount - The amount to subtract from the counter.
     * @returns {Counter} - Chains the instance.
     * @fires set
     */
    decrement(key, amount = 1) {
        const value = Math.floor(amount) || 1;
        return this.set(key, this.get(key) - value);
    }

    /**
     * Removes all counters.
     * @returns {Counter} - Chains the instance.
     * @fires clear
     */
    clear() {
        super.clear();

        this.emit('clear');

        return this;
    }

    /**
     * Loads counter from another source. The counters are cleared before.
     * @param {*} iterator - An iterable object that can be used to set the counters.
     * @returns {Counter} - Chains the instance.
     * @fires load
     */
    load(iterator) {
        if (!iterator || !iterator[Symbol.iterator]) {
            return this;
        }

        super.clear();
        for (const [key, value] of iterator) {
            super.set(key, Math.floor(value) || 0);
        }

        this.emit('load');

        return this;
    }

    /**
     * Exports the counters to an object.
     * @returns {object} - An object containing the counters.
     */
    toObject() {
        const counters = {};

        for (const [key, value] of this.entries()) {
            counters[key] = value;
        }

        return counters;
    }

    /**
     * Validates that the given object is an instance of the class.
     * Otherwise, an error is thrown.
     * @param {object} object - The instance to validate.
     * @throws {TypeError} - If the given object is not a valid instance.
     */
    static validateInstance(object) {
        if (!(object instanceof this)) {
            throw new TypeError(`The object must be an instance of ${this.name}!`);
        }
    }
}

/**
 * Notifies a value has been set to a counter.
 * @event set
 * @param {string} key - The key of the counter that was set.
 * @param {number} newValue - The new value.
 * @param {number} oldValue - The previous value.
 */

/**
 * Notifies a counter has been removed.
 * @event delete
 * @param {string} key - The key of the counter that was removed.
 * @param {number} value - The last value of the removed counter.
 */

/**
 * Notifies all counters has been removed.
 * @event clear
 */

/**
 * Notifies the counters has been loaded.
 * @event load
 */
