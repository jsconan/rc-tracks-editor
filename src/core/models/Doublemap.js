/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { validateCallback } from '../helpers';

/**
 * Defines a doublemap structure.
 *
 * A doublemap is a `Map` where each value is indexed by two keys, a primary and a secondary. Internally, it
 * relies on a combination of `Map` objects. The primary keys are managed by a first `Map`, then each key
 * is associated with a another `Map` object that contains the secondary keys and the related values. In the
 * end, it could be seen as a two-dimensional associative array.
 */
export class Doublemap extends Map {
    /**
     * Creates a doublemap.
     * @param {*} source - A list of primaryKey/secondaryKey/value sets to inject in the map.
     */
    constructor(source) {
        super();

        if (source) {
            for (const [primaryKey, secondaryKey, value] of source) {
                this.set(primaryKey, secondaryKey, value);
            }
        }
    }

    /**
     * The amount of secondary keys registered in the map.
     * @type {number}
     */
    get size2() {
        let size = 0;
        for (const bucket of super.values()) {
            size += bucket.size;
        }
        return size;
    }

    /**
     * Tells whether a key exists or not.
     * It accepts either the primary key only, or the two keys, or the two keys and the value.
     * @param {*} primaryKey - The primary key to check.
     * @param {*} [secondaryKey] - The secondary key to check.
     * @param {*} [value] - An optional value to check together with the pair of keys.
     * @returns {boolean} - Returns `true` if the keys exist.
     */
    has(primaryKey, secondaryKey, value) {
        const undefinedValue = 'undefined' === typeof value;
        if ('undefined' === typeof secondaryKey && undefinedValue) {
            return super.has(primaryKey);
        }

        const bucket = super.get(primaryKey);
        if (!bucket) {
            return false;
        }

        if (undefinedValue) {
            return bucket.has(secondaryKey);
        }

        return bucket.get(secondaryKey) === value;
    }

    /**
     * Gets the value registered under the given pair of keys.
     * If the value does not exist, `undefined` will be returned.
     * @param {*} primaryKey - The primary key to check.
     * @param {*} [secondaryKey] - The secondary key to check.
     * @returns {*} - Returns either the secondary `Map`, the registered value, or `undefined`.
     */
    get(primaryKey, secondaryKey) {
        const bucket = super.get(primaryKey);

        if (!bucket || 'undefined' === typeof secondaryKey) {
            return bucket;
        }

        return bucket.get(secondaryKey);
    }

    /**
     * Sets a value under a pair of keys.
     * @param {*} primaryKey - The primary key to check.
     * @param {*} secondaryKey - The secondary key to check.
     * @param {*} value - The value to register under the keys.
     * @returns {Doublemap} - Chains the instance.
     */
    set(primaryKey, secondaryKey, value) {
        if ('undefined' !== typeof secondaryKey) {
            let bucket = super.get(primaryKey);
            if (!bucket) {
                bucket = new Map();
                super.set(primaryKey, bucket);
            }
            bucket.set(secondaryKey, value);
        }
        return this;
    }

    /**
     * Deletes a key. If only the primary key is supplied, all its secondary keys will be deleted.
     * If all keys are supplied, only the exact pair will be deleted.
     * If the value is also supplied, it will only be removed if it matches the registered value.
     * @param {*} primaryKey - The primary key to check.
     * @param {*} [secondaryKey] - The secondary key to check.
     * @param {*} [value] - An optional value to remove from the key.
     * @returns {boolean} - Returns `true` if something was deleted.
     */
    delete(primaryKey, secondaryKey, value) {
        const undefinedValue = 'undefined' === typeof value;
        if ('undefined' === typeof secondaryKey && undefinedValue) {
            return super.delete(primaryKey);
        }

        const bucket = super.get(primaryKey);
        if (!bucket) {
            return false;
        }

        if (undefinedValue || bucket.get(secondaryKey) === value) {
            const deleted = bucket.delete(secondaryKey);

            if (!bucket.size) {
                super.delete(primaryKey);
            }

            return deleted;
        }

        return false;
    }

    /**
     * Calls a function on every primaryKey/secondaryKey/value sets of the map.
     * @param {eachKeysValue} each - A callback called for each keys/value pair.
     * @param {*} [thisArg] - An optional call context for the callback. If omitted it will be defaulted to `undefined`.
     * @throws {TypeError} - If the callback is not supplied or is not a function.
     */
    forEach(each, thisArg) {
        validateCallback(each);

        for (const [primaryKey, secondaryKey, value] of this.entries()) {
            each.call(thisArg, value, secondaryKey, primaryKey, this);
        }
    }

    /**
     * Produces an iterator to loop over all the keys.
     * If the primary key is supplied, the iterator will list its secondary keys if any.
     * @param {*} [key] - The primary key for which list the secondary keys.
     * @yields {*} - The next value in the map.
     * @generator
     */
    *keys(key) {
        if ('undefined' !== typeof key) {
            const bucket = super.get(key);
            if (bucket) {
                yield* bucket.keys();
            }
            return;
        }

        for (const [primaryKey, bucket] of super.entries()) {
            for (const secondaryKey of bucket.keys()) {
                yield [primaryKey, secondaryKey];
            }
        }
    }

    /**
     * Produces an iterator to loop over all the values.
     * If the primary key is supplied, the iterator will list its values.
     * Otherwise, it will loops through all secondary map as well.
     * @param {*} [key] - The primary key for which list the values.
     * @yields {*} -  The next value in the map.
     * @generator
     */
    *values(key) {
        if ('undefined' !== typeof key) {
            const bucket = super.get(key);
            if (bucket) {
                yield* bucket.values();
            }
            return;
        }

        for (const bucket of super.values()) {
            yield* bucket.values();
        }
    }

    /**
     * Produces an iterator to loop over all the primaryKey/secondaryKey/value sets.
     * If the primary key is supplied, the iterator will list its key/value pairs.
     * @param {*} [key] - The primary key for which list the values.
     * @yields {*} - The next primaryKey/secondaryKey/value set in the map.
     * @generator
     */
    *entries(key) {
        if ('undefined' !== typeof key) {
            const bucket = super.get(key);
            if (bucket) {
                yield* bucket.entries();
            }
            return;
        }

        for (const [primaryKey, bucket] of super.entries()) {
            for (const [secondaryKey, value] of bucket) {
                yield [primaryKey, secondaryKey, value];
            }
        }
    }

    /**
     * Produces an iterator to loop over all the primaryKey/secondaryKey/value sets.
     * @yields {*} - The next primaryKey/secondaryKey/value set in the map.
     * @generator
     */
    *[Symbol.iterator]() {
        yield* this.entries();
    }
}

/**
 * Represents a callback function that is invoked for each primaryKey/secondaryKey/value set of a `Doublemap`
 * object. Mainly used when applying a traversal algorithm, like `forEach()`.
 * @param {*} value - The current value being traversed.
 * @param {*} secondaryKey - The current secondaryKey being traversed.
 * @param {*} primaryKey - The current primaryKey being traversed.
 * @param {Doublemap} map - The `Doublemap` instance being traversed.
 * @callback eachKeysValue
 */
