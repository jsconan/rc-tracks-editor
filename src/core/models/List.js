/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { validateCallback } from '../helpers';
import { eventEmitterMixin } from '../mixins';

/**
 * Represents an observable list.
 * Each change in the list will be notified to the subscribers.
 * @mixes EventEmitter
 */
export class List {
    /**
     * Creates an observable list.
     * Each change in the list will be notified to the subscribers.
     * @param {*} source - An iterable object that can be used to initialize the list.
     */
    constructor(source = []) {
        eventEmitterMixin(this);

        this.list = [...source];
        source = void 0;
    }

    /**
     * The number of values in the list.
     * @type {number}
     */
    get length() {
        return this.list.length;
    }

    /**
     * Iterates over the values from the list.
     * @yields {*} - The next value in the list.
     * @generator
     */
    *[Symbol.iterator]() {
        yield* this.list.values();
    }

    /**
     * Returns an iterator for the values from the list.
     * @yields {*} - The next value in the list.
     * @generator
     */
    *values() {
        yield* this.list.values();
    }

    /**
     * Applies a callback to each value from the list.
     * @param {listCallback} walker - A callback that will be applied to each value of the list.
     * @returns {List} - Chains the list.
     * @throws {TypeError} - If the given callback is not a function.
     */
    forEach(walker) {
        validateCallback(walker);

        this.list.forEach((value, index) => walker.call(this, value, index, this));

        return this;
    }

    /**
     * Maps the values of the list to an array.
     * @param {listCallback} mapper - A mapper callback that will be applied to each value of the list.
     * @returns {Array}
     * @throws {TypeError} - If the given callback is not a function.
     */
    map(mapper) {
        validateCallback(mapper);

        return this.list.map((value, index) => mapper.call(this, value, index, this));
    }

    /**
     * Finds the index of a value from the list. If the values does not exist in the list, `-1` is returned.
     * @param {listCallback|*} filter - Either a filter callback that will be applied to each value of the list
     * and that must return true when the received value matches, or the searched value.
     * @returns {number} - The index of the value in the list, or -1 if not found.
     */
    find(filter) {
        if ('function' === typeof filter) {
            return this.list.findIndex((value, index) => filter.call(this, value, index, this));
        }

        return this.list.indexOf(filter);
    }

    /**
     * Gets the value at a particular index.
     * @param {number} index - The index of the value to get
     * @returns {*} - The value at the given index. It may be undefined.
     */
    get(index) {
        return this.list[index];
    }

    /**
     * Sets a value at a particular index.
     * @param {number} index - The index where to set the value.
     * @param {*} value - The value to set at the index.
     * @returns {List} - Chains the list.
     * @fires set
     * @throws {ReferenceError} - If the given index is out of bounds.
     */
    set(index, value) {
        if (index < 0 || index >= this.length) {
            throw new ReferenceError('The list index is out of bounds!');
        }

        const previous = this.list[index];
        this.list[index] = value;

        /**
         * Notifies a value has been set to the list.
         * @event set
         * @param {number} index - The index where the value was set.
         * @param {*} newValue - The new value.
         * @param {*} oldValue - The previous value.
         */
        this.emit('set', index, value, previous);

        return this;
    }

    /**
     * Inserts a value at a particular index.
     * @param {number} index - The index where to insert the value.
     * @param {...*} value - The value to insert at the index.
     * @returns {List} - Chains the list.
     * @fires insert
     */
    insert(index, ...value) {
        this.list.splice(index, 0, ...value);

        /**
         * Notifies values have been inserted into the list.
         * @event insert
         * @param {number} index - The index where the values have been inserted.
         * @param {...*} value - The inserted values.
         */
        this.emit('insert', index, ...value);

        return this;
    }

    /**
     * Adds a value at the end of the list.
     * @param {...*} value - The value to add.
     * @returns {List} - Chains the list.
     * @fires add
     */
    add(...value) {
        this.list.push(...value);

        /**
         * Notifies values have been added to the list.
         * @event add
         * @param {...*} value - The added value.
         */
        this.emit('add', ...value);

        return this;
    }

    /**
     * Removes a value from the given index.
     * @param {number} index - The index from where remove the value.
     * @param {number} count - The number of values to remove from the index.
     * @returns {number} - The number of deleted values.
     * @fires delete
     */
    delete(index, count = 1) {
        const removed = this.list.splice(index, count);
        const deleted = removed.length;

        if (deleted) {
            /**
             * Notifies values have been removed from the list.
             * @event delete
             * @param {number} index - The index from where the values were removed.
             * @param {...*} value - The removed value.
             */
            this.emit('delete', index, ...removed);
        }

        return deleted;
    }

    /**
     * Clears the list.
     * @returns {List} - Chains the list.
     * @fires clear
     */
    clear() {
        this.list = [];

        /**
         * Notifies the list was cleared.
         * @event clear
         */
        this.emit('clear');

        return this;
    }

    /**
     * Loads values from another source. The list is cleared before.
     * @param {*} iterator - An iterable object that can be used to fill the list.
     * @returns {List} - Chains the list.
     * @fires load
     */
    load(iterator) {
        if (!iterator || !iterator[Symbol.iterator]) {
            return this;
        }

        this.list = [...iterator];

        /**
         * Notifies the list was loaded.
         * @event load
         */
        this.emit('load');

        return this;
    }

    /**
     * Copy the list to an array.
     * @returns {Array} - An array containing the values from the list, in the same order.
     */
    toArray() {
        return [...this.list];
    }

    /**
     * Validates that the given object is an instance of the class.
     * Otherwise, an error is thrown.
     * @param {object} object - The instance to validate.
     * @throws {TypeError} - If the given object is not a valid instance.
     */
    static validateInstance(object) {
        if (!(object instanceof this)) {
            throw new TypeError(`The object must be an instance of ${this.name}!`);
        }
    }
}

/**
 * Callback called from the iteration algorithms.
 * @param {*} value - The current value being traversed.
 * @param {number} index - The index of the current value being traversed.
 * @returns {*} - Returns a value expected by the context.
 * @callback listCallback
 */
