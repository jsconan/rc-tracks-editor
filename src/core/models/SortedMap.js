/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { validateCallback } from '../helpers';
import { TreeNode } from './TreeNode.js';

/**
 * Defines a sorted map structure.
 *
 * A sorted map is a map where the keys are sorted so that a traversal will return a sorted list.
 */
export class SortedMap {
    /**
     * Creates a sorted map.
     * @param {*} source - A list of key/value pairs to inject in the map.
     */
    constructor(source) {
        this.tree = TreeNode.create();
        this.size = 0;

        if (source) {
            for (const [key, value] of source) {
                this.set(key, value);
            }
        }
    }

    /**
     * Tells whether a key exists or not.
     * @param {*} key - The key to check.
     * @returns {boolean} - Returns `true` if the keys exist.
     */
    has(key) {
        return this.tree.has(key);
    }

    /**
     * Gets the value registered under the given key.
     * If the value does not exist, `undefined` will be returned.
     * @param {*} key - The key to check.
     * @returns {*} - Returns the registered value, or `undefined`.
     */
    get(key) {
        return this.tree.get(key);
    }

    /**
     * Sets a value under a key.
     * @param {*} key - The key to set.
     * @param {*} value - The value to register under the key.
     * @returns {SortedMap} - Chains the instance.
     */
    set(key, value) {
        const feedback = {};
        this.tree = this.tree.insert(key, value, feedback);

        if (feedback.added) {
            this.size++;
        }

        return this;
    }

    /**
     * Deletes a key.
     * @param {*} key - The key to remove.
     * @returns {boolean} - Returns `true` if something was deleted.
     */
    delete(key) {
        const feedback = {};
        this.tree = this.tree.delete(key, feedback);

        if (feedback.deleted) {
            this.size--;
            return true;
        }

        return false;
    }

    /**
     * Removes all elements from the map.
     */
    clear() {
        this.tree = TreeNode.create();
        this.size = 0;
    }

    /**
     * Calls a function on every key/value pairs of the map.
     * @param {eachKeyValue} each - A callback called for each keys/value pair.
     * @param {*} [thisArg] - An optional call context for the callback. If omitted it will be defaulted to `undefined`.
     * @throws {TypeError} - If the callback is not supplied or is not a function.
     */
    forEach(each, thisArg) {
        validateCallback(each);

        for (const { key, value } of this.tree) {
            each.call(thisArg, value, key, this);
        }
    }

    /**
     * Produces an iterator to loop over all the keys.
     * @yields {*} - The next key in the map.
     * @generator
     */
    *keys() {
        for (const { key } of this.tree) {
            yield key;
        }
    }

    /**
     * Produces an iterator to loop over all the values.
     * @yields {*} - The next value in the map.
     * @generator
     */
    *values() {
        for (const { value } of this.tree) {
            yield value;
        }
    }

    /**
     * Produces an iterator to loop over all the key/value pairs.
     * @yields {*} - The next key/value pair in the map.
     * @generator
     */
    *entries() {
        for (const { key, value } of this.tree) {
            yield [key, value];
        }
    }

    /**
     * Produces an iterator to loop over all the key/value pairs.
     * @yields {*} - The next key/value pair in the map.
     * @generator
     */
    *[Symbol.iterator]() {
        yield* this.entries();
    }
}

/**
 * Represents a callback function that is invoked for each key/value pair of a `SortedMap` object.
 * Mainly used when applying a traversal algorithm, like `forEach()`.
 * @param {*} value - The current value being traversed.
 * @param {*} key - The current key being traversed.
 * @param {SortedMap} map - The `SortedMap` instance being traversed.
 * @returns {*} - It may return a value that will impact the result of the visit.
 * @callback eachKeyValue
 */
