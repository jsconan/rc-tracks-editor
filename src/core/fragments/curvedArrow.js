/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { enlargeArc, getArcAngle } from '../helpers';
import { SVGPath, Vector2D } from '../models';

/**
 * Builds the SVG path for rendering a curved arrow.
 * @param {number} x - The X-coordinate of the center of the curve.
 * @param {number} y - The Y-coordinate of the center of the curve.
 * @param {number} width - The width of the arrow.
 * @param {number} height - The height of the arrow.
 * @param {number} radius - The radius of the curve.
 * @param {number} angle - Tha angle of the curve.
 * @param {number} thickness - The thickness of the arrow.
 * @param {boolean} clockwise - Is the arrow clockwise or counter-clockwise?
 * @param {number} rotation - The start angle of the curve.
 * @returns {SVGPath} - Returns the path for the shape.
 */
export function curvedArrowPath(
    x,
    y,
    width,
    height,
    radius,
    angle,
    thickness = void 0,
    clockwise = true,
    rotation = 0
) {
    if (!angle) {
        angle = getArcAngle(width, radius);
    }

    if ('number' !== typeof thickness) {
        thickness = height / 3;
    }

    const center = new Vector2D(x, y);
    const tipEdgeRadius = radius;
    const tipInnerRadius = tipEdgeRadius - height / 2;
    const tipOuterRadius = tipInnerRadius + height;
    const innerRadius = tipEdgeRadius - thickness / 2;
    const outerRadius = innerRadius + thickness;
    const bodyAngle = enlargeArc(angle, tipOuterRadius, -height);

    let bodyStart, bodyEnd, tipStart, tipEnd;

    if (clockwise) {
        bodyStart = rotation;
        bodyEnd = bodyStart + bodyAngle;
        tipStart = bodyEnd;
        tipEnd = bodyStart + angle;
    } else {
        bodyEnd = rotation + angle;
        bodyStart = bodyEnd - bodyAngle;
        tipEnd = rotation;
        tipStart = bodyStart;
    }

    const path = new SVGPath();

    if (clockwise) {
        return path
            .polarMoveTo(innerRadius, bodyStart, center)
            .arcCurveTo(innerRadius, -bodyAngle, Vector2D.polar(innerRadius, bodyEnd, center))
            .polarLineTo(tipInnerRadius, tipStart, center)
            .polarLineTo(tipEdgeRadius, tipEnd, center)
            .polarLineTo(tipOuterRadius, tipStart, center)
            .polarLineTo(outerRadius, bodyEnd, center)
            .arcCurveTo(outerRadius, bodyAngle, Vector2D.polar(outerRadius, bodyStart, center))
            .close();
    } else {
        return path
            .polarMoveTo(tipEdgeRadius, tipEnd, center)
            .polarLineTo(tipOuterRadius, tipStart, center)
            .polarLineTo(outerRadius, bodyStart, center)
            .arcCurveTo(outerRadius, -bodyAngle, Vector2D.polar(outerRadius, bodyEnd, center))
            .polarLineTo(innerRadius, bodyEnd, center)
            .arcCurveTo(innerRadius, bodyAngle, Vector2D.polar(innerRadius, bodyStart, center))
            .polarLineTo(tipInnerRadius, tipStart, center)
            .close();
    }
}
