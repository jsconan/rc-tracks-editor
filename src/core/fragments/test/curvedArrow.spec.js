/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { curvedArrowPath } from '../curvedArrow.js';
import { SVGPath } from '../../models';

describe('curvedArrowPath', () => {
    it('is a function', () => {
        expect(curvedArrowPath).toEqual(expect.any(Function));
    });

    it('returns a SVG path', () => {
        expect(curvedArrowPath()).toBeInstanceOf(SVGPath);
    });

    it.each([
        [100, 100, 60, 50, 80, 60, void 0, void 0, void 0],
        [100, 100, 60, 50, 80, 60, 0, void 0, void 0],
        [100, 100, 60, 50, 80, 60, 10, void 0, void 0],
        [100, 100, 60, 50, 80, 60, 10, false, void 0],
        [100, 100, 60, 50, 80, 60, 10, true, void 0],
        [100, 100, 60, 50, 80, 60, 10, false, 30],
        [100, 100, 60, 50, 80, 60, 10, true, 30],
        [100, 100, 60, 50, 80, 360, 10, true, 30],
        [100, 100, 60, 50, 80, 360, 0, true, 30]
    ])(
        'draws a curved arrow tip at %s,%s having width=%s and height=%s, radius=%s and angle=%s, thickness=%s, clockwise=%s, rotated by %s',
        (x, y, width, height, radius, angle, thickness, clockwise, rotation) => {
            expect(
                curvedArrowPath(x, y, width, height, radius, angle, thickness, clockwise, rotation).toString()
            ).toMatchSnapshot();
        }
    );
});
