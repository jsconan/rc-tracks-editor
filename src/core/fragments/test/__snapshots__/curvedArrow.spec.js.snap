// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`curvedArrowPath draws a curved arrow tip at 100,100 having width=60 and height=50, radius=80 and angle=60, thickness=0, clockwise=undefined, rotated by undefined 1`] = `"M 180,100 A 80,80 0 0 1 167.30856787018968,143.23837058983688 L 146.2746404107554,129.72637978051284 L 140,169.28203230275508 L 188.34249532962394,156.7503613991609 L 167.30856787018968,143.23837058983688 A 80,80 0 0 0 180,100 Z"`;

exports[`curvedArrowPath draws a curved arrow tip at 100,100 having width=60 and height=50, radius=80 and angle=60, thickness=10, clockwise=false, rotated by 30 1`] = `"M 169.2820323027551,140 L 156.7503613991609,188.34249532962394 L 145.94076875170168,171.51535336207652 A 85,85 0 0 1 100,185 L 100,175 A 75,75 0 0 0 140.53597242797207,163.1017823783028 L 129.72637978051287,146.2746404107554 Z"`;

exports[`curvedArrowPath draws a curved arrow tip at 100,100 having width=60 and height=50, radius=80 and angle=60, thickness=10, clockwise=false, rotated by undefined 1`] = `"M 180,100 L 193.31850231043313,148.131664489582 L 175.54354948939823,138.9637283963283 A 85,85 0 0 1 142.5,173.61215932167727 L 137.5,164.9519052838329 A 75,75 0 0 0 166.6560730788808,134.37976034970143 L 148.8811202578459,125.21182425644771 Z"`;

exports[`curvedArrowPath draws a curved arrow tip at 100,100 having width=60 and height=50, radius=80 and angle=60, thickness=10, clockwise=true, rotated by 30 1`] = `"M 164.9519052838329,137.5 A 75,75 0 0 1 134.37976034970143,166.6560730788808 L 125.21182425644771,148.8811202578459 L 100,180 L 148.131664489582,193.31850231043313 L 138.9637283963283,175.54354948939823 A 85,85 0 0 0 173.6121593216773,142.5 Z"`;

exports[`curvedArrowPath draws a curved arrow tip at 100,100 having width=60 and height=50, radius=80 and angle=60, thickness=10, clockwise=true, rotated by undefined 1`] = `"M 175,100 A 75,75 0 0 1 163.1017823783028,140.53597242797207 L 146.2746404107554,129.72637978051284 L 140,169.28203230275508 L 188.34249532962394,156.7503613991609 L 171.51535336207652,145.94076875170168 A 85,85 0 0 0 185,100 Z"`;

exports[`curvedArrowPath draws a curved arrow tip at 100,100 having width=60 and height=50, radius=80 and angle=60, thickness=10, clockwise=undefined, rotated by undefined 1`] = `"M 175,100 A 75,75 0 0 1 163.1017823783028,140.53597242797207 L 146.2746404107554,129.72637978051284 L 140,169.28203230275508 L 188.34249532962394,156.7503613991609 L 171.51535336207652,145.94076875170168 A 85,85 0 0 0 185,100 Z"`;

exports[`curvedArrowPath draws a curved arrow tip at 100,100 having width=60 and height=50, radius=80 and angle=60, thickness=undefined, clockwise=undefined, rotated by undefined 1`] = `"M 171.66666666666669,100 A 71.66666666666667,71.66666666666667 0 0 1 160.2972587170449,138.73437365339555 L 146.2746404107554,129.72637978051284 L 140,169.28203230275508 L 188.34249532962394,156.7503613991609 L 174.3198770233344,147.74236752627823 A 88.33333333333334,88.33333333333334 0 0 0 188.33333333333334,100 Z"`;

exports[`curvedArrowPath draws a curved arrow tip at 100,100 having width=60 and height=50, radius=80 and angle=360, thickness=0, clockwise=true, rotated by 30 1`] = `"M 169.2820323027551,140 A 80,80 0 1 1 179.91011496285174,103.79124341394987 L 154.93820403696057,102.60647984709053 L 169.28203230275508,140 L 204.88202588874293,104.9760069808092 L 179.91011496285174,103.79124341394987 A 80,80 0 1 0 169.2820323027551,140 Z"`;

exports[`curvedArrowPath draws a curved arrow tip at 100,100 having width=60 and height=50, radius=80 and angle=360, thickness=10, clockwise=true, rotated by 30 1`] = `"M 164.9519052838329,137.5 A 75,75 0 1 1 174.9157327776735,103.554290700578 L 154.93820403696057,102.60647984709053 L 169.28203230275508,140 L 204.88202588874293,104.9760069808092 L 184.90449714802998,104.02819612732173 A 85,85 0 1 0 173.6121593216773,142.5 Z"`;
